numcodecs/blosc.pyx

```diff
@@ -393,6 +394,99 @@ def decompress(source, dest=None):
     return dest
 
 
+def decompress_partial(source, start, nitems, typesize=0, encoding_size=0, dest=None):
```

Comment:
These are exposed because if the array if encoded to bytes before the buffer is encoded in blosc, then blosc will have 1 as the header size which may not be true.

for example, they typesize and header need to be given to pass this test:
```python
# test encoding of bytes
buf = arr.tobytes(order='A')
enc = codec.encode(buf)
dec = codec.decode_partial(enc, start, nitems, ITEMSIZE, 1)
compare_arrays(compare_arr, dec, precision=precision)
```

I'm open to other ideas for this. Ideally a user wouldn't be using these kwargs unless they knew what they were doing

Reply:
Thanks for clarification. I think the use case for exposing `typesize` in the API is clear. 

Regarding `encoding_size`, is that needed in the API? Shouldn't the `encoding_size` always be the value of the itemsize in the header? Or have I misunderstood? In the tests you are setting it to 1, but that is what would be in the blosc header.

Reply:
I don't think this should be a concern. Zarr tries to maintain the `dtype` of the object being written in part to make sure that Blosc gets the correct `itemsize`. We do this by using a utility function `ensure_ndarray`. So I think we are safe to drop it.

Reply:
@alimanfoo, I made an edit so the encoding size is always determined from from the source buffer header so only the typesize is exposed now.

@jakirkham, If a return buffer isn't given, the typesize is still needed to create it. With the regular decompression, the return buffer can just be created the same size as the source buffer and then `ensure_ndarray` can size it appropriately, but with the partial decompression, the return buffer is sized by `nitems * typesize`. Maybe an option would be to make the destination buffer the same size as the source buffer and then only fill it up to the size of the decompressed data, then just take `slice(0, nitems)`?

Reply:
Sorry let me clarifying.

> These are exposed because if the array if encoded to bytes before the buffer is encoded in blosc, then blosc will have 1 as the header size which may not be true.

What I mean is this case won't happen. A NumPy array with the correct type will be handed to Blosc.

Reply:
Ah so you're saying the cases we have in the tests aren't allowed by zarr?

Reply:
Sorry which test cases are you referring to?

Reply:
[All of these](https://github.com/andrewfulton9/numcodecs/blob/15f482f63a04182a29d4815f500b45a505927a12/numcodecs/tests/common.py#L133-L183). Cases where the array is encoded to bytes, before being encoded by the codec.

Reply:
> [All of these](https://github.com/andrewfulton9/numcodecs/blob/15f482f63a04182a29d4815f500b45a505927a12/numcodecs/tests/common.py#L133-L183). Cases where the array is encoded to bytes, before being encoded by the codec.

Good question. Those tests look like they precede PR ( https://github.com/zarr-developers/numcodecs/pull/128 ), PR ( https://github.com/zarr-developers/zarr-python/pull/352 ), and some other PRs I'm leaving out, which made sure NumPy arrays could be passed through as-is.

@alimanfoo, are we missing anything about these tests?

Reply:
@jakirkham and @alimanfoo, Any more thoughts about this? If the encoding is controlled by zarr, then I think it makes sense to leave out the `typesize` option. If we do that. I'll also remove the tests with pre-encoded arrays as well.

Reply:
SGTM. Let's see what Alistair says :)

Reply:
Hi folks, sorry for slow reply. 

FWIW I think you can assume that the typesize has been correctly passed through to blosc during write, and therefore that it can always be read from the header.

All of the tests with different non-numpy array-like things were originally added to ensure that the decode and encode methods would accept any object exposing a (new-style) buffer interface. Of course in this situation it causes a problem because the conversion e.g. from numpy to bytes causes a loss of information about the typesize. So FWIW I think you could either (a) drop the tests involving things which don't propagate the original typesize, or (b) adjust those tests to compare against the expectation given the new typesize after conversion from numpy to something else.

Reply:
To elaborate a little, numcodecs is intended to be used by zarr but also other libraries. The tests ensure that numcodecs can handle as input any object which exposes the new-style buffer interface.

However, numcodecs can also assume that whatever typesize ([itemsize](https://docs.python.org/3/c-api/buffer.html#c.Py_buffer.itemsize)) is exposed via that buffer interface is the right one. 

I.e., if a user has performed some conversion of their data upstream of numcodecs which has caused some loss or change to the typesize (itemsize) that numcodecs receives, that is the user's problem.

Reply:
encoding size is now removed and I updated the tests to handle it appropriately